### PR TITLE
Add some missing barriers when dealing with global caches

### PR DESCRIPTION
**Pull request by jaykrell**
accessed from multiple threads without locks.

This is based on earlier PR https://github.com/mono/mono/pull/17849..

Read-side barriers are removed.
  More analysis is needed to determine if they are needed (which pointer
  read vs. written).
Write-side barriers are store-release or store-store barrier instead of full barrier (which separately will get compiler barriers).

Note all candidates are converted to new form yet (go slowly).

### CODE REVIEW

File: mono/metadata/class-internals.h [hunk 1]
```diff
@@ -1507,6 +1518,38 @@ mono_method_is_constructor (MonoMethod *method);
...
+	type *name; 								\
+	name = static_ ## name;							\
+	if (!name) {								\
+		/* Custom code here to initialize name */
+#define MONO_STATIC_POINTER_INIT_END(type, name)				\
+		if (name) {							\
+			/* Success, commit to static. */			\
+			mono_atomic_store_seq (&static_ ## name, name);		\
+		} else {							\
+			/* Try one more time in case another thread did it. */	\
```
Comment (BrzVlad) [inline mono/metadata/class-internals.h]: I'm not convinced this case is needed, or whether it guarantees anything.

Comment (jaykrell) [reply mono/metadata/class-internals.h]: The “try one more time” part? Ok, will remove.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

File: mono/metadata/class-internals.h [hunk 2]
```diff
@@ -1507,6 +1518,35 @@ mono_method_is_constructor (MonoMethod *method);
...
+#define MONO_STATIC_POINTER_INIT(type, name)					\
+	static type *static_ ## name;						\
+	type *name; 								\
+	name = static_ ## name;							\
+	if (!name) {								\
+		/* Custom code here to initialize name */
+#define MONO_STATIC_POINTER_INIT_END(type, name)				\
+		if (name) {							\
+			/* Success, commit to static. */			\
+			mono_atomic_store_seq (&static_ ## name, name);		\
```
Comment (jaykrell) [inline mono/metadata/class-internals.h]: This line `mono_atomic_store_seq` might be stronger than desired.

File: mono/metadata/class-internals.h [hunk 3]
```diff
@@ -1507,6 +1518,35 @@ mono_method_is_constructor (MonoMethod *method);
...
+// It may be desirable to replace this with mono_lazy_initialize, etc.
+//
+// These macros cannot be wrapped in do/while as they inject "name" into invoking scope.
+//
+#define MONO_STATIC_POINTER_INIT(type, name)					\
+	static type *static_ ## name;						\
+	type *name; 								\
+	name = static_ ## name;							\
+	if (!name) {								\
+		/* Custom code here to initialize name */
```
Comment (jaykrell) [inline mono/metadata/class-internals.h]: Is this formatting, of the two macros right next to each other, with a comment, useful or confusing?
I was of course trying to be informative, but when I read it, I find it confusing.
Maybe something more verbose, like "user puts custom code between the macros", which really is saying the same thing.

**Comment by jaykrell** [comment]
@monojenkins squash

**Pull request merged**